### PR TITLE
[qa] Refactor test_framework and pull tester

### PR DESCRIPTION
**Pull request by maflcko**
- log to stdout
- increase range for p2p and rpc ports
- UPPERCASE_CONSTANTS
- Stop nodes on CTRL+C

### CODE REVIEW

**Comment by MarcoFalke** [comment]
* Added a commit to properly stop other nodes, even when one fails to stop
* Added a commit to update the Readme

**Comment by MarcoFalke** [comment]
Anyone interested in looking at the diff here? (It is mostly move-only and cleaning up the logic/ifelses)

It is already tested as part of #7972 but I'd prefer to have at least one read the patch.

File: qa/rpc-tests/test_framework/util.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 from base64 import b64encode
 from decimal import Decimal, ROUND_DOWN
 import json
+import httplib
```
Comment (MarcoFalke) [inline qa/rpc-tests/test_framework/util.py]: Needs bump: https://github.com/bitcoin/bitcoin/pull/7972#issuecomment-217399645

File: qa/rpc-tests/test_framework/util.py [hunk 2]
```diff
@@ -27,6 +28,13 @@
 
 COVERAGE_DIR = None
 
+# The maximum number of nodes a single test can spawn
+MAX_NODES = 8
+# Don't assign rpc or p2p ports lower than this
+PORT_MIN = 11000
+# The number of ports to "reserve" for p2p and rpc, each
+PORT_RANGE = 5000
```
Comment (MarcoFalke) [inline qa/rpc-tests/test_framework/util.py]: @theuni I am assuming this works fine on travis?

File: qa/rpc-tests/test_framework/util.py (hunk 1 above)
Comment (MarcoFalke) [inline qa/rpc-tests/test_framework/util.py]: @fanquake @paveljanik done in another commit.

**Comment by MarcoFalke** [comment]
If someone wants to test this, make sure to merge it with current bitcoin/master

**Comment by paveljanik** [comment]
`README.md` should state install python3-zmq.

**Comment by MarcoFalke** [comment]
@paveljanik  It is already updated in master. Sorry, I didn't want to rebase this pull. (Can you do the rebase/merge locally?)

File: qa/rpc-tests/test_framework/util.py (hunk 1 above)
Comment (MarcoFalke) [inline qa/rpc-tests/test_framework/util.py]: Rebased instead :stuck_out_tongue_winking_eye:

**Comment by MarcoFalke** [comment]
@paveljanik Rebased, so the other doc changes are included as well.

**Comment by paveljanik** [comment]
I have a lot of these during tests, but not only in the PR - also applies to current master:
```
Running testscript wallet.py ...
Initializing test directory /var/folders/65/fn0h49r55k7779vg1b_h461r0000gn/T/testbv4onukz
Unexpected exception caught during testing: timeout('timed out',)
...
Running testscript abandonconflict.py ...
Initializing test directory /var/folders/65/fn0h49r55k7779vg1b_h461r0000gn/T/test4arfwesr
Unexpected exception caught during testing: timeout('timed out',)
```

**Comment by MarcoFalke** [comment]
It's just refactoring. Ideally, behavior should stay the same.

File: qa/pull-tester/rpc-tests.py [hunk 3]
```diff
@@ -153,48 +161,35 @@
...
+    elif len(opts) == 0 or (len(opts) == 1 and "-win" in opts):
+        test_list = testScripts
+    else:
+        for t in testScripts + testScriptsExt:
+            if t in opts or re.sub(".py$", "", t) in opts:
+                test_list.append(t)
+
+    if print_help:
+        # Help should be the same for all scripts, so just
+        # call the first and exit
```
Comment (sdaftuar) [inline qa/pull-tester/rpc-tests.py]: comment ultranit: the help is not actually the same for all scripts, as some tests take custom arguments.

**Comment by sdaftuar** [comment]
utACK (feel free to ignore my ultranit)

Not sure this needs to be addressed in this PR, but for supporting multiple tests running in parallel, I think it might make sense to add support to the test framework for specifying a range of ports to use for the p2p and rpc ports on the command line, which the `rpc_tests.py` script could use.  The logic now of hoping that pid's don't collide modulo some big number (which is now bigger after this PR) is not sufficiently robust in my opinion.  Doesn't need to be fixed here, just mentioning it since this PR is touching those functions already.

File: qa/pull-tester/rpc-tests.py (hunk 3 above)
Comment (MarcoFalke) [inline qa/pull-tester/rpc-tests.py]: This is true, but the previous behavior was to only show the help text of the first script and then `exit(0)`. If someone wants to change the behavior, I'd propose a separate pull. I am happy to change the help text in this pull, though.

**Comment by MarcoFalke** [comment]
Added a commit to fix @sdaftuar's nit, hopefully. In regard to the ports: I am wondering if it is sufficient to pass in a int `seed (default=pid)` instead of trying to parse ranges from the command line.

**Comment by sdaftuar** [comment]
@marcofalke Yeah that approach of a single int that defaults to the pid sounds reasonable to me. Comment looks better too, thanks.
